### PR TITLE
Remove WebAssemblyLoggerFactory and refactor logging setup

### PR DESCRIPTION
**Pull request by captainsafia**
**Changes in this PR**
 - Replaces `WebAssemblyLoggerFactory` with `LoggerFactory` from logging extensions package
 - Moves WebAssemblyConsoleLogger and PrependMessageLogger to provider model

Now that we are using the standard `LoggerFactory` support for config options like `SetMinimumLevel` and `AddProvider` is available.

Compared to what is currently in the `blazor-wasm` branch, the changes in this PR add an additional 12 kb to the total compressed size.

Addresses #19737

### CODE REVIEW

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:26]:
```suggestion
        private readonly List<Action<IServiceCollection>> _configServices = new List<Action<IServiceCollection>>();
```

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs:20]: Couldn't this be written without introducing a new extension method - hostBuilder.Services.AddLogging(...)`?

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:17]: This is perfectly fine. If you wanted to, you could probably get away with using a vanilla `Dictionary` since the environment is single threaded.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:18]:
```suggestion
        private readonly IJSInProcessRuntime _jsRuntime;
```

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:17]:
```suggestion
        private readonly ConcurrentDictionary<string, WebAssemblyConsoleLogger<object>> _loggers;
```

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:42]: Hmm, this is a strange. We generally don't null out fields after use. Is there a good reason to do this?

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:172]: I'm not super sure we need this indirection. If you have access to the builder, you should be able to just add services to it.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (javiercn) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:42]: It's true that we don't, but there's also no harm in it tbh, specially since this is a singleton

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs
Comment (javiercn) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs:20]: I would say this is nice in the sense that it keeps consistency with the ASP.NET Core pattern about it, but I don't feel super strongly about it. That said, it's a single method so I think we should not be concerned. Do you have any other concerns in mind?

**Review by javiercn — COMMENTED** [review]
Looks great, the only thing I think it's missing is an e2e test that uses it in conjunction with configuration (as it's the most common pattern in ASP.NET Core).

I would also suggest we consider updating the project templates to align them with what we do in other ASP.NET Core scenarios with regards to the login level (put login level in appsettings.json and appsettings.development.json) and have those two included in the default template.

**Comment by SteveSandersonMS** [comment]
> I would also suggest we consider updating the project templates to align them with what we do in other ASP.NET Core scenarios with regards to the login level (put login level in appsettings.json and appsettings.development.json) and have those two included in the default template.

I see the argument towards consistency and discoverability, but on the other hand, it's an extra HTTP request at runtime on every load for people who aren't otherwise using config. If we think most people won't customize the logging level (and I don't), then I lean towards having the default be ">= Warning" and not having config files in the template.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs
Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs:20]: If possible, @pranavkm's suggestion (`hostBuilder.Services.AddLogging`) keeps us closer aligned to the simplified startup model that @rynowak was working us towards.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:15]: Could you rename the file to match the new type name? Thanks!

Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:15]: Also, does this need to be public? If we can make it `internal` that would be great.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:17]: @pranavkm We're typically trying to account for thread safety in the usual multi-threaded style now whenever possible, because pretty soon we'll want to enable the threaded version of Mono WebAssembly. At that point we'll need to go through the code and hunt down any thread safety issues due to older assumptions.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:42]: If we are nulling them out, could do so unconditionally without tracking `_disposed`

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:49]: Oh I see - there are two top-level classes in this file. Would it be possible to split them into different files? It's just to make it quicker to locate the file containing a given class, e.g., when searching by filename on GitHub.

File: src/Components/test/testassets/BasicTestApp/PrependMessageLoggerProvider.cs
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/PrependMessageLoggerProvider.cs:12]: Great!

Comment (SteveSandersonMS) [reply src/Components/test/testassets/BasicTestApp/PrependMessageLoggerProvider.cs:12]: BTW super minor nitpick, but this doesn't need to be disposable.

**Comment by SteveSandersonMS** [comment]
This looks superb to me. I'm really glad you noticed we could switch to the standard factory without a major size penalty.

I agree with @javiercn that it would be good to see an E2E test in here too. However if this risks missing the preview 4 date then I'd be totally happy for that test to be added as a follow-up.

One question: does `WebAssemblyConsoleLogger`'s `IsEnabled` logic need to be updated so it's not hardcoded to "warning" level any more? I'm not sure how configuring the level in a config file would be able to work if this is hardcoded. Or if that's meant to be out of scope for this PR that's fine (but we should follow up later).

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilderExtensions.cs:20]: Yeah, to clarify:

`hostBuilder.ConfigureLogging` -> good
`hostBuilder.ConfigureServices` -> meh

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:17]: Oh fun. That makes `WebAssemblyConfigurationBuilder : IConfigurationBuilder, IConfiguration` a tad bit trickier.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:42]: > but there's also no harm in it tbh,

You can no longer make fields `readonly` which results in a poorer pattern. Removing it seems fine.

Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:42]: Ok yes, agreed.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (captainsafia) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:18]: This value is modified in the Dispose method so it's not immutable.

Comment (captainsafia) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:18]: Saw the thread below. Makes sense. Updated.

File: src/Components/test/testassets/BasicTestApp/PrependMessageLoggerProvider.cs
Comment (captainsafia) [reply src/Components/test/testassets/BasicTestApp/PrependMessageLoggerProvider.cs:12]: Correct. ILoggerProvider extends IDisposable. Fixed.

**Comment by captainsafia** [comment]
Thanks for the feedback everyone, @SteveSandersonMS, @pranavkm, @javiercn . I've pushed a commit with changes. Here's a summary of what was done.

- Fixed minor comments around internal vs.public classes and file organization
- Removed the ConfigureLogging extension method in favor of directly exposing the LoggingBuilder on the WebAssemblyHostBuilder

> One question: does WebAssemblyConsoleLogger's IsEnabled logic need to be updated so it's not hardcoded to "warning" level any more? I'm not sure how configuring the level in a config file would be able to work if this is hardcoded. Or if that's meant to be out of scope for this PR that's fine (but we should follow up later).

Yes. That's my bad. I've removed the `>= LogLevel.Warning` check within the console longer. Now this is handled by the LoggerFactory itself.

> I would also suggest we consider updating the project templates to align them with what we do in other ASP.NET Core scenarios with regards to the login level (put login level in appsettings.json and appsettings.development.json) and have those two included in the default template.

So, this ends up being a little bit tricky. Adding configuration options to the LoggingBuilder requires having access to an IConfiguration object. However, within our builder we only have access to the IConfigurationBuilder. As a result, the developer does not have access to an IConfiguration object when they are setting up their DI.

```
var builder = WebAssemblyHostBuilder.CreateDefault(args);
builder.Logging.AddConfiguration() // No IConfiguration object here


var configuration = builder.Configuration.Build();
builder.Logging.AddConfiguration(configuration); // Causes exception documented in issue below
```

This problem is similar to https://github.com/dotnet/aspnetcore/issues/20231. IMO, we should take a look making the Configuration object that we expose adhere to the IConfiguration interface so that it can be used during startup.

Thoughts?

**Comment by SteveSandersonMS** [comment]
>  Adding configuration options to the LoggingBuilder requires having access to an IConfiguration object ...
> we should take a look at making the Configuration object that we expose adhere to the IConfiguration interface so that we can read the config without having to build it.

Agreed.

So are you saying that if we did fix our `builder.Configuration` implementation so that it was safe to read out `IConfiguration` instances multiple times, then this would automatically work with `builder.Logging.AddConfiguration(configuration)`? If that's the case, then I think it can be considered an external issue to this PR and there's nothing else you need to do for it. We should fix that underlying issue separately, and then we'll get the benefits here without further changes.

On the other hand, if you're saying it still wouldn't work for some reason even if you could safely get an `IConfiguration` instance, could you clarify why? Thanks!

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:60]: If I'm following correctly, this is defining a default, but the developer could still override it by calling `builder.Logging.SetMinimumLevel(somethingElse)` afterwards? If so that's ideal!

**Review by SteveSandersonMS — APPROVED** [review]
This looks awesome!

Approving now, but one final request if you think there's time. In this part of the E2E tests:

https://github.com/dotnet/aspnetcore/blob/blazor-wasm/src/Components/test/E2ETest/Tests/WebAssemblyLoggingTest.cs#L83-L89

... what do you think about [amending the test app](https://github.com/dotnet/aspnetcore/blob/blazor-wasm/src/Components/test/testassets/BasicTestApp/Program.cs#L47-L55) so that its log level is more detailed, and then having it test that it can log at the other levels (< warning) too?

If you think this would make the console output too noisy when working on other non-logging-related tests, then fine not to do this. I'm not sure there's any way to do it selectively just for this one test, unless you spot anything.

**Comment by captainsafia** [comment]
> ... what do you think about amending the test app so that its log level is more detailed, and then having it test that it can log at the other levels (< warning) too?

Pushed a commit with this change.

I was trying very hard to get this to work by reading a config specific to the PrependMessage provider but ran into the issues above. Later, I'd like to go back and re-write this test once we've added support for `Logging.AddConfiguration`.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs
Comment (captainsafia) [reply src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:60]: Yes, this is possible. Based on the recommendation below, I updated BasicTestApp to decrease the min level.

**Review by javiercn — APPROVED** [review]
Looks fantastic!
Thanks for following through!

Just for sanity, once we fix the config thing, do we need to do anything for logging to pick the config? If that is the case, can we file a separate issue to track that. Or add a comment on the existing issue regarding making config work from startup?

**Comment by captainsafia** [comment]
> Just for sanity, once we fix the config thing, do we need to do anything for logging to pick the config? If that is the case, can we file a separate issue to track that. Or add a comment on the existing issue regarding making config work from startup?

We shouldn't need to do anything. The developer will have to call `Logging.AddConfiguration` in their startup code and pass it the new `wasmHostBuilder.Configuration` IConfiguration object that we will support.

**Pull request merged**
